### PR TITLE
Add implementation and contract change for serialization surrogate

### LINKED ISSUE #1931: Add support for data contract surrogate type aliasing

### PR DESCRIPTION
**Pull request by khdang**
Fix #1931

### CODE REVIEW

**Comment by khdang** [comment]
cc @SGuyGe @shmao @roncain

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 1]
```diff
@@ -1846,6 +1846,66 @@ public static void DCS_DeserializeEmptyString()
         });
     }
 
+    [Fact]
+    public static void DCS_DataContractSurrogate()
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Rename it to DCS_MyPersonSurrogate() to be more specific?

File: src/System.Runtime.Serialization.Xml/ref/project.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
     "System.Runtime": "4.0.0",
     "System.Xml.ReaderWriter": "4.0.0",
     "System.IO": "4.0.0",
-    "System.Runtime.Serialization.Primitives": "4.0.10",
+    "System.Runtime.Serialization.Primitives": "4.1.0-beta-23423",
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/ref/project.json]: Do we want to reference this specific version of the package? Should this be "4.1.0-beta-*"?

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs (hunk 1 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Done

File: src/System.Runtime.Serialization.Xml/ref/project.json (hunk 2 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/ref/project.json]: We shouldn't have reference to this specific version. I've updated as suggestion.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs [hunk 3]
```diff
@@ -21,13 +21,15 @@ public class XmlObjectSerializerWriteContextComplex : XmlObjectSerializerWriteCo
     internal class XmlObjectSerializerWriteContextComplex : XmlObjectSerializerWriteContext
 #endif
     {
+        protected ISerializationSurrogateProvider SerializationSurrogateProvider;
         private SerializationMode _mode;
-
+
```
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs]: nit: remove extra spaces.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerReadContextComplex.cs [hunk 4]
```diff
@@ -26,12 +26,14 @@ internal class XmlObjectSerializerReadContextComplex : XmlObjectSerializerReadCo
 
         private bool _preserveObjectReferences;
         private SerializationMode _mode;
+        protected ISerializationSurrogateProvider SerializationSurrogateProvider;
```
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerReadContextComplex.cs]: Question: do we need this field to be "protected"?

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs [hunk 5]
```diff
@@ -21,13 +21,15 @@ public class XmlObjectSerializerWriteContextComplex : XmlObjectSerializerWriteCo
     internal class XmlObjectSerializerWriteContextComplex : XmlObjectSerializerWriteContext
 #endif
     {
+        protected ISerializationSurrogateProvider SerializationSurrogateProvider;
```
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs]: Question: do we need this field to be "protected"?

**Comment by shmao** [comment]
:shipit: :+1:

**Comment by SGuyGe** [comment]
LGTM

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerReadContextComplex.cs (hunk 4 above)
Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerReadContextComplex.cs]: For now it doesn't need to be protected. I was thinking of adding the surrogate support for DCJS as well which would need this to be protected but we've agreed that can be done in a separated  change.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs (hunk 5 above)
Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs]: Same comment as for XmlObjectSerializerReadContextComplex.cs

**Comment by zhenlan** [comment]
LGTM

**Comment by stephentoub** [comment]
@dotnet-bot test this please (https://github.com/dotnet/corefx/issues/4172)

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs [hunk 6]
```diff
@@ -448,5 +458,23 @@ internal override Type GetDeserializeType()
...
+
+        internal static object SurrogateToDataContractType(ISerializationSurrogateProvider serializationSurrogateProvider, object oldObj, Type surrogatedDeclaredType, ref Type objType)
+        {
+            object obj = DataContractSurrogateCaller.GetObjectToSerialize(serializationSurrogateProvider, oldObj, objType, surrogatedDeclaredType);
+            if (obj != oldObj)
+            {
+                if (obj == null)
+                    objType = Globals.TypeOfObject;
+                else
+                    objType = obj.GetType();
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs]: Nit: could be simplified to:
```C#
objType = obj != null ? obj.GetType() : Globals.TypeOfObject;
```

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSurrogateCaller.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    static class DataContractSurrogateCaller
+    {
+        internal static Type GetDataContractType(ISerializationSurrogateProvider surrogateProvider, Type type)
+        {
+            if (DataContract.GetBuiltInDataContract(type) != null)
+                return type;
+            Type dcType = surrogateProvider.GetSurrogateType(type);
+            if (dcType == null)
+                return type;
+            return dcType;
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSurrogateCaller.cs]: Nit: previous four lines could be simplified to:
```C#
return surrogateProvider.GetSurrogateType(type) ?? type;
```

File: src/System.Runtime.Serialization.Xml/ref/project.json [hunk 8]
```diff
@@ -3,7 +3,7 @@
     "System.Runtime": "4.0.0",
     "System.Xml.ReaderWriter": "4.0.0",
     "System.IO": "4.0.0",
-    "System.Runtime.Serialization.Primitives": "4.0.10",
+    "System.Runtime.Serialization.Primitives": "4.1.0-beta-*",
```
Comment (stephentoub) [inline src/System.Runtime.Serialization.Xml/ref/project.json]: It's ok to update this dependency?

**Comment by stephentoub** [comment]
A few minor comments, but otherwise LGTM.

File: src/System.Runtime.Serialization.Xml/ref/project.json (hunk 8 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/ref/project.json]: We've added a new interface to System.Runtime.Serialization.Primitives contract. From discussion we've agreed to take dependency of the new S.R.S.Primitives package.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs (hunk 6 above)
Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs]: Updated as suggestion

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSurrogateCaller.cs (hunk 7 above)
Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSurrogateCaller.cs]: Updated as suggestion

**Comment by khdang** [comment]
@dotnet-bot test this please

**Comment by khdang** [comment]
The CodeFormatter check failures are:
```
11:14:52 Formatting native file with command: clang-format -style=file d:\j\workspace\dotnet_corefx\native_code_format_check_prtest\src\Native\System.Native\pal_errno.cpp
11:14:52 File not style compliant, exiting: System.Native\pal_errno.cpp
11:14:52 
11:14:52 d:\j\workspace\dotnet_corefx\native_code_format_check_prtest>exit -1 
11:14:52 Build step 'Execute Windows batch command' marked build as failure
```
Hi @stephentoub My understanding is that we want to merge this soon so that it can be integrate to rc1 branch in time. Is there anything we can do about the failure? My first commit before addressing the comments and squash passed all the tests, should we go ahead and merge this change?

**Comment by stephentoub** [comment]
You're not changing native code.  Don't worry about the Code Formatter Check.  If it passes the main leg with all the builds and tests, you're welcome to merge.  We're working on getting the code formatter leg fixed.

**Pull request merged**